crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -1071,16 +863,65 @@ pub(crate) unsafe fn get_mut_by_id(
     entity: Entity,
     location: EntityLocation,
     component_id: ComponentId,
-) -> Option<MutUntyped> {
+) -> Option<MutUntyped<'_>> {
     let change_tick = world.change_tick();
+    // SAFETY: component_id is valid
     let info = world.components.get_info_unchecked(component_id);
-    // SAFETY: world access is unique, entity location and component_id required to be valid
-    get_component_and_ticks(world, component_id, info.storage_type(), entity, location).map(
-        |(value, ticks)| MutUntyped {
+    // SAFETY:
+    // - world access is unique
+    // - entity location is valid
+    // - and returned component is of type T
+    world
+        .get_component_and_ticks(component_id, info.storage_type(), entity, location)
+        .map(|(value, ticks)| MutUntyped {
+            // SAFETY: world access is unique and ties world lifetime to `MutUntyped` lifetime
             value: value.assert_unique(),
             ticks: Ticks::from_tick_cells(ticks, world.last_change_tick(), change_tick),
-        },
-    )
+        })
+}
+
+/// Moves component data out of storage.
+///
+/// This function leaves the underlying memory unchanged, but the component behind
+/// returned pointer is semantically owned by the caller and will not be dropped in its original location.
+/// Caller is responsible to drop component data behind returned pointer.
+///
+/// # Safety
+/// - `entity_location` must be within bounds of the given archetype and `entity` must exist inside the `archetype`
+/// - `component_id` must be valid
+/// - `components`, `archetypes` and `storages` must come from the same world
+/// - The relevant table row **must be removed** by the caller once all components are taken
+#[inline]
+pub(crate) unsafe fn take_component<'a>(
```

Comment:
The PR description suggests that this is a method on Storages (which I like better), but here it's a freestanding function.